File: src/hooks/task-completed.ts
import {
  createTask,
  extractCwd,
  isConfigured,
  loadHookConfig,
  postMemoryEvent,
  readHookInput,
  updateTaskStatus,
  writeHookOutput,
} from "./common.js";

function firstString(input: Record<string, unknown>, keys: string[]): string | undefined {
  for (const key of keys) {
    const value = input[key];
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  const task = input.task;
  if (task && typeof task === "object") {
    return firstString(task as Record<string, unknown>, keys);
  }
  return undefined;
}

function looksLikeRecoveryTask(description: string): boolean {
  const lower = description.toLowerCase();
  return ["error", "failure", "retry", "recover", "fix", "incident"].some((keyword) =>
    lower.includes(keyword)
  );
}

export async function runTaskCompletedHook(): Promise<void> {
  const input = readHookInput<Record<string, unknown>>();
  const cwd = extractCwd(input);
  const config = loadHookConfig(cwd);

  const taskId = firstString(input, ["task_id", "taskId"]) || "";
  const taskSubject =
    firstString(input, ["task_subject", "task_title", "title", "subject", "description"]) ||
    "Completed task";
  const taskDescription = firstString(input, ["task_description", "details", "content"]);
  const planId = firstString(input, ["plan_id", "planId"]);

  if (process.env.CONTEXTSTREAM_TASK_COMPLETED_REQUIRE_SUBJECT === "true" && !taskSubject.trim()) {
    writeHookOutput({ blocked: true, reason: "TaskCompleted requires a non-empty task subject" });
    return;
  }

  if (isConfigured(config)) {
    let updated = false;
    if (taskId) {
      updated = await updateTaskStatus(
        config,
        taskId,
        "completed",
        taskSubject,
        taskDescription
      );
    }

    if (!updated) {
      await createTask(config, {
        title: taskSubject,
        description: taskDescription,
        planId: planId || undefined,
        status: "completed",
      });
    }

    await postMemoryEvent(
      config,
      "Task completed",
      {
        task_id: taskId || null,
        title: taskSubject,
        description: taskDescription,
        plan_id: planId || null,
        agent_id: firstString(input, ["agent_id", "agentId"]) || null,
        team_name: firstString(input, ["team_name", "teamName"]) || null,
        completed_at: new Date().toISOString(),
        source: "task_completed_hook",
      },
      ["hook", "task", "completed"],
      "task"
    ).catch(() => {});

    if (taskDescription && looksLikeRecoveryTask(taskDescription)) {
      await postMemoryEvent(
        config,
        "Lesson from task completion",
        {
          task: taskSubject,
          description: taskDescription,
          lesson: "Recovered from an execution issue; consider codifying this into tests/guards.",
        },
        ["hook", "lesson", "task_completed"],
        "lesson"
      ).catch(() => {});
    }
  }

  writeHookOutput();
}

const isDirectRun =
  process.argv[1]?.includes("task-completed") || process.argv[2] === "task-completed";
if (isDirectRun) {
  runTaskCompletedHook().catch(() => process.exit(0));
}

File: crates/mcp-server/src/hook_handlers/task_completed.rs
//! TaskCompleted hook handler.
//!
//! Captures task completion events into ContextStream tasks and memory events.

use anyhow::Result;
use serde_json::Value;

use super::common::{create_task, extract_cwd, load_config, post_memory_event, update_task_status};
use super::{read_stdin_json, write_stdout_json, HookOutput};

/// Handle the TaskCompleted hook.
pub async fn handle() -> Result<()> {
    let input = read_stdin_json()?;
    let cwd = extract_cwd(&input);
    let config = load_config(&cwd);

    let task_id = first_str(&input, &["task_id", "taskId"]).unwrap_or("");
    let task_subject = first_str(
        &input,
        &[
            "task_subject",
            "task_title",
            "title",
            "subject",
            "description",
        ],
    )
    .unwrap_or("Completed task")
    .trim()
    .to_string();
    let task_description = first_str(&input, &["task_description", "details", "content"])
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(String::from);
    let plan_id = first_str(&input, &["plan_id", "planId"])
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(String::from);

    // Optional gate: block completion if task subject is missing.
    let require_subject = std::env::var("CONTEXTSTREAM_TASK_COMPLETED_REQUIRE_SUBJECT")
        .map(|v| v == "true")
        .unwrap_or(false);
    if require_subject && task_subject.is_empty() {
        eprintln!("TaskCompleted requires a non-empty task subject");
        std::process::exit(2);
    }

    if config.is_configured() {
        let mut updated = false;
        if !task_id.is_empty() {
            updated = update_task_status(
                &config,
                task_id,
                "completed",
                Some(&task_subject),
                task_description.as_deref(),
            )
            .await;
        }

        if !updated {
            let _ = create_task(
                &config,
                &task_subject,
                task_description.as_deref(),
                plan_id.as_deref(),
                Some("completed"),
            )
            .await;
        }

        let content = serde_json::json!({
            "task_id": if task_id.is_empty() { Value::Null } else { Value::String(task_id.to_string()) },
            "title": task_subject,
            "description": task_description,
            "plan_id": plan_id,
            "agent_id": first_str(&input, &["agent_id", "agentId"]),
            "teammate_name": first_str(&input, &["teammate_name", "teammateName"]),
            "team_name": first_str(&input, &["team_name", "teamName"]),
            "completed_at": chrono::Utc::now().to_rfc3339(),
            "source": "task_completed_hook"
        });

        post_memory_event(
            &config,
            "Task completed",
            content,
            &["hook", "task", "completed"],
        )
        .await;

        if looks_like_recovery_task(task_description.as_deref().unwrap_or("")) {
            let lesson = serde_json::json!({
                "task": task_subject,
                "description": task_description,
                "lesson": "Recovered from an execution issue; consider codifying this into tests/guards.",
            });
            post_memory_event(
                &config,
                "Lesson from task completion",
                lesson,
                &["hook", "lesson", "task_completed"],
            )
            .await;
        }
    }

    write_stdout_json(&HookOutput::empty())?;
    Ok(())
}

fn first_str<'a>(value: &'a Value, keys: &[&str]) -> Option<&'a str> {
    keys.iter()
        .find_map(|k| value.get(*k).and_then(|v| v.as_str()))
        .or_else(|| {
            value.get("task").and_then(|task| {
                keys.iter()
                    .find_map(|k| task.get(*k).and_then(|v| v.as_str()))
            })
        })
}

fn looks_like_recovery_task(description: &str) -> bool {
    if description.is_empty() {
        return false;
    }
    let lower = description.to_lowercase();
    ["error", "failure", "retry", "recover", "fix", "incident"]
        .iter()
        .any(|keyword| lower.contains(keyword))
}
